File: app/src/main/java/com/suzei/racoon/ui/add/AddSingleChatFragment.java
package com.suzei.racoon.ui.add;


import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ProgressBar;

import android.widget.TextView;

import com.suzei.racoon.R;
import com.suzei.racoon.ui.add.search.SearchAdapter;
import com.suzei.racoon.model.Users;
import com.suzei.racoon.ui.chatroom.single.SingleChatActivity;
import com.suzei.racoon.ui.add.search.SearchContract;
import com.suzei.racoon.ui.add.search.SearchPresenter;
import com.suzei.racoon.view.EmptyRecyclerView;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class AddSingleChatFragment extends Fragment implements SearchContract.SearchView {

    private SearchPresenter searchPresenter;

    @BindView(R.id.pick_user_no_result) TextView noResultView;
    @BindView(R.id.pick_user_loading) ProgressBar progressBarView;
    @BindView(R.id.pick_user_back) ImageButton backView;
    @BindView(R.id.pick_user_search_text) EditText searchUserView;
    @BindView(R.id.pick_user_search) ImageButton searchView;
    @BindView(R.id.pick_user_search_list) EmptyRecyclerView listSearchUserView;

    public AddSingleChatFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.pick_user, container, false);
        initObjects(view);
        setUpRecyclerViews();
        return view;
    }

    private void initObjects(View view) {
        ButterKnife.bind(this, view);
        searchPresenter = new SearchPresenter(this);
    }

    private void setUpRecyclerViews() {
        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext(),
                LinearLayoutManager.VERTICAL, false);
        listSearchUserView.setLayoutManager(layoutManager);
        listSearchUserView.setItemAnimator(new DefaultItemAnimator());
        listSearchUserView.setEmptyView(noResultView);
    }

    @OnClick(R.id.pick_user_back)
    public void onBackClick() {
        getActivity().finish();
    }

    @OnClick(R.id.pick_user_search)
    public void onSearchClick() {
        ArrayList<Users> users = new ArrayList<>();

        String query = searchUserView.getText().toString().trim();
        searchPresenter.startSearch(query, users);
    }

    @Override
    public void setSearchAdapter(SearchAdapter searchAdapter) {
        listSearchUserView.setAdapter(searchAdapter);
    }

    @Override
    public void setSearchUserItemClick(Users users) {
        Intent chatRoomIntent = new Intent(getContext(), SingleChatActivity.class);
        chatRoomIntent.putExtra(SingleChatActivity.EXTRA_ID, users.getUid());
        startActivity(chatRoomIntent);
        getActivity().finish();
    }

    @Override
    public void searchSuccess() {
    }

    @Override
    public void searchFailed() {

    }

    @Override
    public void showProgress() {
        searchView.setEnabled(false);
        progressBarView.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideProgress() {
        searchView.setEnabled(true);
        progressBarView.setVisibility(View.GONE);
    }
}

File: app/src/main/java/com/suzei/racoon/ui/add/AddSingleChatFragment.kt
package com.suzei.racoon.ui.add

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import com.suzei.racoon.databinding.PickUserBinding
import com.suzei.racoon.model.Users
import com.suzei.racoon.ui.add.search.SearchAdapter
import com.suzei.racoon.ui.add.search.SearchContract
import com.suzei.racoon.ui.add.search.SearchPresenter
import com.suzei.racoon.ui.chatroom.single.SingleChatActivity
import java.util.*

class AddSingleChatFragment : Fragment(), SearchContract.SearchView {
    private var searchPresenter: SearchPresenter? = null

    private var _binding: PickUserBinding? = null
    private val binding get() = _binding!!

    private val onBackClick = View.OnClickListener {
        requireActivity().finish()
    }

    private val onSearchClick = View.OnClickListener {
        val users = ArrayList<Users>()
        val query = binding.pickUserSearchText.text.toString().trim { it <= ' ' }
        searchPresenter!!.startSearch(query, users)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        _binding = PickUserBinding.inflate(layoutInflater)
        initObjects()
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setUpRecyclerViews()
        setupClickListener()
    }

    override fun onDestroy() {
        super.onDestroy()
        _binding = null
    }

    private fun initObjects() {
        searchPresenter = SearchPresenter(this)
    }

    private fun setUpRecyclerViews() {
        val layoutManager = LinearLayoutManager(
            context,
            LinearLayoutManager.VERTICAL, false
        )

        binding.pickUserSearchList.layoutManager = layoutManager
        binding.pickUserSearchList.itemAnimator = DefaultItemAnimator()
        binding.pickUserSearchList.setEmptyView(binding.pickUserNoResult)
    }

    private fun setupClickListener() {
        binding.pickUserBack.setOnClickListener(onBackClick)
        binding.pickUserSearch.setOnClickListener(onSearchClick)
    }

    override fun setSearchAdapter(searchAdapter: SearchAdapter) {
        binding.pickUserSearchList.adapter = searchAdapter
    }

    override fun setSearchUserItemClick(users: Users) {
        val chatRoomIntent = Intent(context, SingleChatActivity::class.java)
        chatRoomIntent.putExtra(SingleChatActivity.EXTRA_ID, users.uid)
        startActivity(chatRoomIntent)
        requireActivity().finish()
    }

    override fun searchSuccess() {}
    override fun searchFailed() {}
    override fun showProgress() {
        binding.pickUserSearch.isEnabled = false
        binding.pickUserLoading.visibility = View.VISIBLE
    }

    override fun hideProgress() {
        binding.pickUserSearch.isEnabled = true
        binding.pickUserLoading.visibility = View.GONE
    }
}